File: decorator/kotlin/ComporessionDecorator.kt
package Decorator

// Yeni davranışların tanımlandığı sınıftır.
// UML diyagramındaki BaseDecorator (BaseDataSourceDecorator) sınıfından türer.
// UML diyagramındaki ConcreteDecorator sınıfına denk gelmektedir.

internal class ComporessionDecorator(dataSource: IDataSource) : BaseDataSourceDecorator(dataSource) {

    override val fileName: String
        get() = super.dataSource.fileName

    override fun readData() {
        // Veriyi okurken referansını tuttuğumuz sınıfın kendi davranışını sergilemesini istiyoruz.
        super.dataSource.readData()
    }

    override fun writeData(data: Any) {
        // Burada gelen veriye ait sıkıştırma işlemleri yapılacaktır.
        println("Data compressed.")

        // Sıkıştırma işlemleri yapıldıktan sonra yazma işlemleri gerçekleştirilecektir.
        println("Compressed data was written to " + super.dataSource.fileName)
    }
}

File: decorator/go/comporession-decorator.go
package main

import "fmt"

// Yeni davranışların tanımlandığı sınıftır.
// UML diyagramındaki ConcreteDecorator sınıfına denk gelmektedir.
type compressionDecorator struct {
	ds DataSource
}

func NewCompressionDecorator(ds DataSource) *compressionDecorator {
	return &compressionDecorator{ds: ds}
}

func (c compressionDecorator) ReadData() {
	// Veriyi okurken referansını tuttuğumuz struct'ın kendi davranışını sergilemesini istiyoruz.
	c.ds.ReadData()
}

func (c *compressionDecorator) WriteData(data []byte) {
	// Burada gelen veriye ait sıkıştırma işlemleri yapılacaktır.
	fmt.Println("Data compressed.")

	// Sıkıştırma işlemleri yapıldıktan sonra yazma işlemleri gerçekleştirilecektir.
	fmt.Println("Compressed data was written to " + c.ds.GetFileName())
}
